src/SourceBuild/content/test/Microsoft.DotNet.SourceBuild.SmokeTests/SourcelinkTests.cs

```diff
@@ -0,0 +1,136 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.DotNet.SourceBuild.SmokeTests;
+
+public class SourcelinkTests : SmokeTests
+{
+    private static string SourcelinkRoot { get; } = Path.Combine(Directory.GetCurrentDirectory(), "sourcelink");
+
+    public SourcelinkTests(ITestOutputHelper outputHelper) : base(outputHelper) { }
+
+    /// <summary>
+    /// Verifies that all symbols have valid sourcelinks.
+    /// </summary>
+    [SkippableFact(Config.ExcludeSourcelinkEnv, skipOnTrueEnv: true)]
+    public void VerifySourcelinks()
+    {
+        if (Directory.Exists(SourcelinkRoot))
+        {
+            Directory.Delete(SourcelinkRoot, true);
+        }
+        Directory.CreateDirectory(SourcelinkRoot);
+
+        IList<string> failedFiles = ValidateAllFiles(ExtractPackages(GetAllSymbolsPackages()), GetSourcelinkToolPath());
+
+        foreach (string file in failedFiles)
+        {
+            OutputHelper.WriteLine($"Failed sourcelink verification: {file}");
+        }
+
+        Assert.True(failedFiles.Count == 0);
+    }
+
+    /// <summary>
+    /// Gets the path to sourcelink binary.
+    /// Extracts the dotnet-sourcelink tool package from PSB arhive.
+    /// </summary>
+    /// <returns>Path to sourcelink tool binary.</returns>
+    private string GetSourcelinkToolPath()
+    {
+        const string SourcelinkToolPackageNamePattern = "dotnet-sourcelink*nupkg";
+        const string SourcelinkToolBinaryFilenamePattern = "dotnet-sourcelink.dll";
+
+        string toolPackageDir = Directory.CreateDirectory(Path.Combine(SourcelinkRoot, "sourcelink-tool")).FullName;
+        Utilities.ExtractTarball(Config.SourceBuiltArtifactsPath, toolPackageDir, SourcelinkToolPackageNamePattern);
+        string[] files = Directory.GetFiles(toolPackageDir, SourcelinkToolPackageNamePattern, SearchOption.AllDirectories);
```

Comment:
This pattern appears to be used three times.  Consider making a shared Utility method for it.

Reply:
Quoted code shows tarball extraction which appears twice. Are you suggesting a Utility method for the pattern that follows the extraction - getting a single file and failing if there are 0 or more than 1?

Reply:
Sorry my comment was not very clear.  I am referring to this pattern:

```
        files = Directory.GetFiles(extractedToolPath, SourcelinkToolBinaryFilenamePattern, SearchOption.AllDirectories);
        Assert.False(files.Length > 1, $"Unexpected - found more than one sourcelink tool binary with filename pattern: {SourcelinkToolBinaryFilenamePattern}");
        Assert.False(files.Length == 0, $"Did not find sourcelink tool binary with expected filename pattern: {SourcelinkToolBinaryFilenamePattern}");
```

Reply:
I'm not sure about this. One of the instances is going away in the near future, when we start using Symbols archives. The code gets harder to read, line lengths balloon, or line breaks get introduced that just invite bugs.

Reply:
For clarity this is what I mean.  Usage `Utility.GetFile(path, pattern)`

public static string GetFile(string path, string pattern)
{
        files = Directory.GetFiles(path, pattern, SearchOption.AllDirectories);
        Assert.False(files.Length > 1, $"Found multiple files matching the pattern `{pattern}`: {Enivronment.NewLine}{string.Join(files, Enivronment.NewLine)}");
        Assert.False(files.Length == 0, $"Did not find any files matching the pattern `{pattern}`");
        
        return file[0];
}

To me it actually helps the readability of the code because it simplifies this method.

Reply:
Yeah, that would be a generic simplification,. I wanted to provide a way to supply, and use, custom messages, which bloated the code in the caller method, and made it almost unreadable.

I'll do the suggested solution.